hbase-server/src/main/java/org/apache/hadoop/hbase/master/MetaFixer.java

```diff
@@ -244,4 +301,47 @@ static boolean isOverlap(RegionInfo ri, Pair<RegionInfo, RegionInfo> pair) {
     }
     return ri.isOverlap(pair.getFirst()) || ri.isOverlap(pair.getSecond());
   }
+
+  /**
+   * A union over {@link L} and {@link R}.
+   */
+  private static class Either<L, R> {
+    private final L left;
+    private final R right;
+
+    public static <L, R> Either<L, R> ofLeft(L left) {
+      return new Either<>(left, null);
+    }
+
+    public static <L, R> Either<L, R> ofRight(R right) {
+      return new Either<>(null, right);
+    }
+
+    Either(L left, R right) {
+      this.left = left;
+      this.right = right;
+    }
+
+    public boolean hasLeft() {
+      return left != null;
+    }
+
+    public L getLeft() {
+      if (!hasLeft()) {
+        throw new IllegalStateException("Either contains no left.");
+      }
+      return left;
+    }
+
+    public boolean hasRight() {
+      return right != null;
+    }
+
+    public R getRight() {
+      if (!hasRight()) {
+        throw new IllegalStateException("Either contains no right.");
+      }
+      return right;
+    }
+  }
 }
```

Comment:
Fancy. As long as unit tests still pass.... (I think the unit test was pretty involved IIRC)